### PR TITLE
Analyses using functional (FC) and structural connectivity (SC)

### PR DESCRIPTION
**Pull request by man-shu**
This PR aims to achieve the following:

- [x] Classify subjects and runs using FC (measured as Pearson's correlations, tangent space embeddings) from resting state and movie-watching sessions
- [x] Correlate FC vs FC and FC vs SC for - all pairs of ROIs, ROIs within each hemisphere, and ROIs across hemispheres.
- [x] Compare the FC vs FC and FC vs SC correlations within vs across subjects (to see whether these correlations are subject-specific) 
- [x] Investigate which specific ROIs are more important for this subject-specificity

### CODE REVIEW

**Comment by man-shu** [comment]
Hi @bthirion, I have pushed an initial draft for the classification task in [`scripts/connectivity_func_classifier.py`](https://github.com/individual-brain-charting/public_analysis_code/blob/connectivity_analysis/scripts/connectivity_func_classifier.py). Right now, I am classifying subjects (or runs) in two ways:

1. With runs (or subjects) as features: max 18 runs (or 12 subjects)
2. With connectivity between pairs of ROIs as features: ~ 80000 ROIs

I noticed in the [nilearn example](https://nilearn.github.io/stable/auto_examples/03_connectivity/plot_group_level_connectivity.html) they use the 2nd approach with ~700 ROIs. For us, it gives 100% accuracy for subject classification and 0% for run classification. With the 1st approach, the accuracy is at chance level in both cases.

Let me know what you think.

Thanks!

File: ibc_public/utils_connectivity.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        for subject_session in subject_sessions:
+            if (
+                subject_session[0] in ["sub-11", "sub-12"]
+                and subject_session[1] == "ses-13"
+            ):
+                continue
+            try:
+                sub_ses[subject_session[0]]
+            except KeyError:
+                sub_ses[subject_session[0]] = []
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:65]: Why do we need this ?

File: ibc_public/utils_connectivity.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            sub_ses[subject_session[0]].append(subject_session[1])
+    return sub_ses, modality
+
+
+def get_all_subject_connectivity(
+    sub_ses,
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:71]: 'subject_session' rather than 'sub_ses'

File: ibc_public/utils_connectivity.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    all_subject_connectivity = []
+
+    for sub, sess in sub_ses.items():
+        if conn_measure.split('_')[0] == "Pearsons" or conn_measure == "TangentSpaceEmbedding":
+            sess = sorted(sess)
+            subject_connectivity = []
+            subject_connectivity_ = []
+            for i, ses in enumerate(sess):
+                if conn_measure == "TangentSpaceEmbedding":
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:108]: Why do we need that ?

File: scripts/connectivity_func_classifier.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+from sklearn.model_selection import StratifiedShuffleSplit
+from sklearn.svm import LinearSVC
+from ibc_public.utils_connectivity import get_ses_modality, get_all_subject_connectivity
+import itertools
+import os
+import pandas as pd
+from joblib import Parallel, delayed
+
+
+def classify(task, connectivity_measures, scenarios, features, DATA_ROOT, cv_splits=5, cv_test_size=0.25, out_dir="func_classification"):
```
Comment (bthirion) [inline scripts/connectivity_func_classifier.py:12]:
```suggestion
def classify(
     task, 
     connectivity_measures, 
     scenarios, 
     features, 
     DATA_ROOT, 
     cv_splits=5, 
     cv_test_size=0.25, 
     out_dir="func_classification"):
```

File: scripts/connectivity_func_classifier.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                print(feature)
+                ss[conn_measure][f"{scenario}_{feature}"] = []
+                # store the data, cross-validation, classes and scores
+                if scenario == "subs":
+                    if feature == "subsruns":
+                        # print(NotImplementedError)
+                        # continue
+                        # reshape the data to have 
+                        # dim 1 (n_samples) = n_subs * n_rois_per_run 
+                        # and dim 2 (n_features) = n_runs_per_sub
```
Comment (bthirion) [inline scripts/connectivity_func_classifier.py:57]: What is dim 0 ?

File: scripts/connectivity_func_classifier.py (hunk 4 above)
Comment (bthirion) [reply scripts/connectivity_func_classifier.py:12]: what are 'scenarios' ?

File: ibc_public/utils_connectivity.py (hunk 1 above)
Comment (man-shu) [reply ibc_public/utils_connectivity.py:73]: I added some comments in the script to clarify this better

File: ibc_public/utils_connectivity.py [hunk 6]
```diff
@@ -75,82 +213,546 @@ def get_ses_modality(task):
...
+):
+    """Cross validate the given connectomes using the given classes.
+
+    Parameters
+    ----------
+    connectomes : numpy array
+        Array containing the connectomes to classify. Each row is a vectorised connectome.
+    classes : numpy array
+        Array containing the class for each connectome.
+    splits : numpy array
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py]: or list of arrays ? Can you maybe clarif the expected structure, array dimension, etc.

File: ibc_public/utils_connectivity.py [hunk 7]
```diff
@@ -75,82 +213,546 @@ def get_ses_modality(task):
...
+            n_splits = len(task_df)
+            for _, row in task_df.iterrows():
+                classes.extend(task_df["true_class"].tolist())
+                classes.extend(task_df["predicted_class"].tolist())
+            classes = np.concatenate(classes)
+            classes = np.asarray(classes)
+            classes = np.unique(classes)
+            task_df["labels"] = [classes for split in range(n_splits)]
+            task_df["n_labels"] = [len(classes) for split in range(n_splits)]
+            task_df["chance"] = [1 / len(classes) for split in range(n_splits)]
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py]: you assume that the classes are balanced... which may not be the case.

Comment (bthirion) [reply ibc_public/utils_connectivity.py]: Maybe better to rely on a DummyClassifier

File: ibc_public/utils_connectivity.py [hunk 8]
```diff
@@ -0,0 +1,790 @@
...
+        results_dir,
+        f"{task_label}_{connectivity_measure}_results.csv",
+    )
+    results_df.to_csv(results_file)
+    # do_plots(results, results_dir, classify)
+
+    return results_df
+
+
+def do_plots(all_results, results_dir):
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:537]: If you take it as a function of the public API, please add a docstring

File: ibc_public/utils_connectivity.py [hunk 9]
```diff
@@ -0,0 +1,790 @@
...
+                    1 / len(classes) for split in range(n_splits)
+                ]
+                dfs.append(task_df)
+    df = pd.concat(dfs)
+    df["cov"] = df["connectivity"].str.split(" ").str[0]
+    df["cor"] = df["connectivity"].str.split(" ").str[1]
+    return df
+
+
+def plot_accuracy(all_results, classes, cov, results_dir):
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:601]: Same here.

File: ibc_public/utils_connectivity.py [hunk 10]
```diff
@@ -0,0 +1,790 @@
...
+            text.set_visible(False)
+            handle.set_visible(False)
+    legend.set_title("Task")
+    plt.xlabel("FC measure")
+    plt.ylabel("Accuracy")
+    plt.savefig(plot_file, bbox_inches="tight", transparent=True)
+    plt.close()
+
+
+def plot_auc(all_results, classes, cov, results_dir):
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:659]: And here

File: ibc_public/utils_connectivity.py [hunk 11]
```diff
@@ -0,0 +1,790 @@
...
+        if i > 3:
+            text.set_visible(False)
+            handle.set_visible(False)
+    plt.xlabel("FC measure")
+    plt.ylabel("AUC")
+    plt.savefig(plot_file, bbox_inches="tight", transparent=True)
+    plt.close()
+
+
+def plot_confusion(df, classes, task, connectivity_measure, results_dir):
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:694]: Here

File: ibc_public/utils_connectivity.py [hunk 12]
```diff
@@ -0,0 +1,790 @@
...
+    )
+    plot_file = os.path.join(
+        confusion_dir,
+        f"{classes}_{task}_{connectivity_measure}_confusion.png",
+    )
+    plt.savefig(plot_file, bbox_inches="tight")
+    plt.close()
+
+
+def plot_cv_indices(
```
Comment (bthirion) [inline ibc_public/utils_connectivity.py:722]: needs to be public ?

File: scripts/connectivity_func_v_struc.py [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+    "GroupSparseCovarianceCV_corr",
+    "GroupSparseCovarianceCV_partcorr",
+]
+# calculated measures of structural connectivity
+STRUC_CONN_MEASURES = [
+    "_connectome_schaefer400_MNI152_siftweighted",
+]
+
+
+def select_rois(connectivities, n_rois, selection_criterion):
```
Comment (bthirion) [inline scripts/connectivity_func_v_struc.py:28]: Please include at least a 1-line docstring to help understanding the code.

Comment (bthirion) [reply scripts/connectivity_func_v_struc.py:28]: (and same for functions below)

File: scripts/connectivity_func_v_struc.py [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+        )
+        plot_file = os.path.join(
+            boxplot_dir, f"withinvacross_FCSC_{row['selected_rois']}_box.png"
+        )
+        plt.savefig(plot_file, bbox_inches="tight")
+        plt.close()
+
+    # plot boxplot for FC-SC correlation but compare across different roi selections
+    # one plot for within subject and one for across subject
+    d = {
```
Comment (bthirion) [inline scripts/connectivity_func_v_struc.py:448]: Can you find a better name. using single letters that doe not correspond to a clear standard makes it error prone and hard to debug.

**Review by bthirion — COMMENTED** [review]
Most of my comments are about docstrings. This takes time but helps a lot in the long term...

File: ibc_public/utils_connectivity.py (hunk 7 above)
Comment (man-shu) [reply ibc_public/utils_connectivity.py]: Ok yes! That makes sense, I'll do that

**Comment by bthirion** [comment]
Do you need another review on that one ?

**Comment by man-shu** [comment]
I just updated everything with my local changes. 

The conflict is just the session file that I had to update manually to get Mario's sessions, and it can be resolved easily.

We could merge it here, but do you think that would be sufficient to go with the paper, too? 
Or would we have to make a separate repo for that?

**Comment by bthirion** [comment]
We should merge this. We may not need an additional repository.

**Review by bthirion — COMMENTED** [review]
Please remove the code that is commented if it's no longer used.

**Comment by man-shu** [comment]
Yeah the code is super messy, and haven't even pushed all the scripts yet.
I am going to reorganise it under a separate `connectivity` module today.

**Comment by man-shu** [comment]
I have now created separate module `connectivity` under `ibc_public` with separate `utils_*` for plotting, FC estimation, FC classification etc..

Also created a sub-directory `connectivity` under `scripts` with shorter and more specific scripts.

I know this would be difficult to review but all the connectivity related stuff is separate and would not intervene with previous code in the package. It is just for the analyses/figures in the connectivity paper.

**Comment by bthirion** [comment]
Is it ready for review now ?

**Comment by man-shu** [comment]
Yes it is now!

File: scripts/connectivity/plotting/plot_all_accuracy_table [hunk 15]
```diff
@@ -0,0 +1,79 @@
+import os
```
Comment (bthirion) [inline scripts/connectivity/plotting/plot_all_accuracy_table:1]: Rename to plot_all_accuracy_table.py

File: scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
+"""Pipeline to calculate functional connectivity and fit classifier for
```
Comment (bthirion) [inline scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py:1]: wim being the first name of a person, I suggest to change the name.
Same thing for next script.

File: scripts/connectivity/supplementary/generalize_wim_to_ibc.py [hunk 17]
```diff
@@ -0,0 +1,293 @@
+"""This script trains a classifier on Wim (or Mantini et al.)
```
Comment (bthirion) [inline scripts/connectivity/supplementary/generalize_wim_to_ibc.py:1]: Same thing here

**Review by bthirion — COMMENTED** [review]
Minor renaming. Can be merged otherwise.

File: scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py (hunk 16 above)
Comment (man-shu) [reply scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py:1]: ok yes I will change it to the `mantini`

Comment (bthirion) [reply scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py:1]: Mantini is also a name. I'd avoid this.

Comment (man-shu) [reply scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py:1]: Does `external` sound ok?

Comment (bthirion) [reply scripts/connectivity/supplementary/estimate_fc_ibc_sync_wim.py:1]: Yes.

**Comment by man-shu** [comment]
Done! Merging...
